src/Microsoft.Extensions.SecretManager.Tools/Internal/ProjectIdResolver.cs

```diff
@@ -0,0 +1,129 @@
+        private Stack<Action> _disposables = new Stack<Action>();
+
+        public ProjectIdResolver(ILogger logger, string workingDirectory)
+        {
+            _workingDirectory = workingDirectory;
+            _logger = logger;
+        }
+
+        public string Resolve(string project, string configuration = Constants.DefaultConfiguration)
+        {
+            var finder = new MsBuildProjectFinder(_workingDirectory);
+            var projectFile = finder.FindMsBuildProject(project);
+
+            _logger.LogDebug(Resources.Message_Project_File_Path, projectFile);
+
+            var targetFile = GetTargetFile();
+            var outputFile = Path.GetTempFileName();
+            _disposables.Push(() => TryDelete(outputFile));
+
+            var commandOutput = new List<string>();
+            var commandResult = Command.CreateDotNet("msbuild",
+                new[] {
+                    targetFile,
+                    "/nologo",
+                    "/t:_FindUserSecretsProperty",
+                    $"/p:Project={projectFile}",
+                    $"/p:OutputFile={outputFile}",
+                    $"/p:Configuration={configuration}"
+                })
+                .CaptureStdErr()
+                .CaptureStdOut()
+                .OnErrorLine(l => commandOutput.Add(l))
+                .OnOutputLine(l => commandOutput.Add(l))
+                .Execute();
+
+            if (commandResult.ExitCode != 0)
+            {
+                _logger.LogDebug(string.Join(Environment.NewLine, commandOutput));
+                throw new GracefulException(Resources.FormatError_ProjectFailedToLoad(projectFile));
+            }
+
+            var id = File.ReadAllText(outputFile);
+            if (string.IsNullOrEmpty(id))
+            {
+                throw new GracefulException(Resources.FormatError_ProjectMissingId(projectFile));
+            }
+
+            return id.Trim();
+        }
+
+        public void Dispose()
+        {
+            while (_disposables.Count > 0)
+            {
+                _disposables.Pop().Invoke();
+            }
+        }
+
+        private string GetTargetFile()
+        {
+            var assemblyDir = Path.GetDirectoryName(GetType().GetTypeInfo().Assembly.Location);
+
+            // targets should be in one of these locations, depending on test setup and tools installation
+            var searchPaths = new[]
+            {
+                AppContext.BaseDirectory,
+                assemblyDir, // next to assembly
+                Path.Combine(assemblyDir, "../../tools"), // inside the nupkg
+            };
+
+            var foundFile = searchPaths
+                .Select(dir => Path.Combine(dir, TargetsFileName))
+                .Where(File.Exists)
+                .FirstOrDefault();
+
+            if (foundFile != null)
+            {
+                return foundFile;
+            }
+
+            // back up plan - if the targets file was not found any of the expected locations,
```

Comment:
Since this is used today for testing/non prod scenarios make that clear in the comment for future devs